File: proxy_lib/src/main/java/wxgaly/android/proxy_lib/ProxyApplication.java
package wxgaly.android.proxy_lib;

import android.app.Application;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FilenameFilter;
import java.io.IOException;
import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * wxgaly.android.proxy_lib.
 * <p>
 * 加固的代理Application，目前由于打包出的jar中只有工程的类，如果使用kotlin的话就缺少kotlin的依赖jar，
 * 所以使用java实现，也是为了减少jar包的大小，加快安装速度
 *
 * @author Created by WXG on 2019-11-01 11:09.
 * @version V1.0
 */
public class ProxyApplication extends Application {

    private static final String TAG = "ProxyApplication";

    /**
     * 定义好的加密后的文件的存放路径
     */
    private String appName;
    private String appVersion;
    private boolean isBindReal;
    private Application delegate;


    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        Log.d(TAG, "attachBaseContext");

        //获取用户填入的metaData
        getMetaData();

        //准备dex文件
        if (Contants.BUILD_TYPE.equalsIgnoreCase(BuildConfig.BUILD_TYPE)) {
            prepareDexFiles();
        }

    }

    /**
     * 开始替换application
     */
    @Override
    public void onCreate() {
        super.onCreate();
        try {
            bindRealApplication();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 让代码走入if的第三段中
     *
     * @return
     */
    @Override
    public String getPackageName() {
        if (!TextUtils.isEmpty(appName)) {
            return "";
        }
        return super.getPackageName();
    }

    @Override
    public Context createPackageContext(String packageName, int flags) throws PackageManager.NameNotFoundException {
        if (TextUtils.isEmpty(appName)) {
            return super.createPackageContext(packageName, flags);
        }
        try {
            bindRealApplication();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return delegate;

    }

    private void getMetaData() {
        try {
            ApplicationInfo applicationInfo = getPackageManager().getApplicationInfo(
                    getPackageName(), PackageManager.GET_META_DATA);
            Bundle metaData = applicationInfo.metaData;
            if (null != metaData) {
                if (metaData.containsKey(Contants.APP_NAME)) {
                    appName = metaData.getString(Contants.APP_NAME);
                }
                if (metaData.containsKey(Contants.APP_VERSION)) {
                    appVersion = metaData.getString(Contants.APP_VERSION);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 准备dex文件，如果解压过并解密过则直接加载。
     */
    private void prepareDexFiles() {
        Log.d(TAG, "prepare dex files");
        //得到当前apk文件
        File apkFile = new File(getApplicationInfo().sourceDir);

        //把apk解压  这个目录中的内容需要root权限才能使用
        File appDexDir = getDir(appName + "_" + this.appVersion, MODE_PRIVATE);

//        File appDir = new File(appDexDir, "app");
        File dexDir = new File(appDexDir, "dexDir");

        //得到我们需要加载的dex文件
        List<File> dexFiles = new ArrayList<>();
        //进行解密 （最好做md5文件校验）
        if (!dexDir.exists() || dexDir.list().length == 0) {
            if (!dexDir.exists()) {
                dexDir.mkdirs();
            }
            //把apk解压到appDir
            try {
                List<String> strings = ZipUtil.GetFileNameList(apkFile.getAbsolutePath());
                List<String> dexsFileNames = new ArrayList<>();
                for (String string : strings) {
                    if (string.endsWith(Contants.DEXS_SUFFIX)) {
                        Log.d(TAG, "zip: " + string);
                        dexsFileNames.add(string);
                    }
                }
                ZipUtil.UnZipFolder(apkFile.getAbsolutePath(), dexDir.getAbsolutePath(), dexsFileNames);
            } catch (Exception e) {
                e.printStackTrace();
            }
            //获取目录下所有的文件
            File[] files = dexDir.listFiles();
            for (File file : files) {
                String name = file.getName();
                if (name.endsWith(Contants.DEXS_SUFFIX) && !TextUtils.equals(name, "classes.dex")) {
                    try {
                        DesCryptUtil desCryptUtil = new DesCryptUtil(Contants.DEFAULT_KEY);
                        //读取文件内容
                        byte[] bytes = FileUtil.readFileBytes(file);
                        //解密
                        byte[] decrypt = desCryptUtil.decrypt(bytes);
                        //写到指定的目录
                        File decryptDex = new File(dexDir, name);
                        FileOutputStream fos = new FileOutputStream(decryptDex);
                        fos.write(decrypt);
                        fos.flush();
                        fos.close();
                        Log.d(TAG, "decryptDex: " + decryptDex.getAbsolutePath());
                        dexFiles.add(decryptDex);

                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        } else {
            Collections.addAll(dexFiles, dexDir.listFiles(new FilenameFilter() {
                @Override
                public boolean accept(File dir, String name) {
                    return name.endsWith(Contants.DEXS_SUFFIX);
                }
            }));
        }

        try {
            loadDex(dexFiles, dexDir);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 该方法的兼容性要注意，
     * 5.1以下的时候，方法名为makePathElements
     * 5.1以上包括5.1的时候，方法名为makeDexElements
     * 8.1该方法报错，目前没有查看8.1源码，不知道是否有其他变动
     *
     * @param dexFiles
     * @param dexDir
     * @throws Exception
     */
    private void loadDex(List<File> dexFiles, File dexDir) throws Exception {
        //1、获取pathList
        Field pathListField = ReflectUtil.getDeclaredField(getClassLoader(), "pathList");
        pathListField.setAccessible(true);
        Object pathList = pathListField.get(getClassLoader());
        //2、获取数组dexElements
        Field dexElementsField = ReflectUtil.getDeclaredField(pathList, "dexElements");
        dexElementsField.setAccessible(true);
        Object[] dexElements = (Object[]) dexElementsField.get(pathList);
        //3、反射到初始化makeDexElements的方法（注：Android5.1以后，Google对方法名进行了修改，原方法名为makePathElements，参数类型是(List.class, File.class, List.class)）
        Method makeDexElements = ReflectUtil.getMethod(pathList, "makeDexElements", ArrayList.class, File.class, ArrayList.class);
        makeDexElements.setAccessible(true);
        ArrayList<IOException> suppressedException = new ArrayList<>();
        Object[] addElements = (Object[]) makeDexElements.invoke(pathList, dexFiles, dexDir, suppressedException);

        Object[] newElements = (Object[]) Array.newInstance(dexElements.getClass().getComponentType(), dexElements.length + addElements.length);
        System.arraycopy(dexElements, 0, newElements, 0, dexElements.length);
        System.arraycopy(addElements, 0, newElements, dexElements.length, addElements.length);

        //替换classloader中的element数组
        dexElementsField.set(pathList, newElements);
    }

    /**
     * 下面主要是通过反射系统源码的内容，然后进行处理，把我们的内容加进去处理
     */
    private void bindRealApplication() throws Exception {
        if (isBindReal) {
            return;
        }

        if (TextUtils.isEmpty(appName)) {
            return;
        }

        //得到attchBaseContext(context) 传入的上下文 ContextImpl
        Context baseContext = getBaseContext();
        //创建用户真实的application  （MyApplication）
        Class<?> delegateClass = null;
        delegateClass = Class.forName(appName);

        delegate = (Application) delegateClass.newInstance();

        //得到attch()方法
        Method attach = Application.class.getDeclaredMethod("attach", Context.class);
        attach.setAccessible(true);
        attach.invoke(delegate, baseContext);

        //获取ContextImpl ----> ,mOuterContext(app);  通过Application的attachBaseContext回调参数获取
        Class<?> contextImplClass = Class.forName("android.app.ContextImpl");
        //获取mOuterContext属性
        Field mOuterContextField = contextImplClass.getDeclaredField("mOuterContext");
        mOuterContextField.setAccessible(true);
        mOuterContextField.set(baseContext, delegate);

        //ActivityThread  ----> mAllApplication(ArrayList)  ContextImpl的mMainThread属性
        Field mMainThreadField = contextImplClass.getDeclaredField("mMainThread");
        mMainThreadField.setAccessible(true);
        Object mMainThread = mMainThreadField.get(baseContext);

        //ActivityThread  ----->  mInitialApplication       ContextImpl的mMainThread属性
        Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
        Field mInitialApplicationField = activityThreadClass.getDeclaredField("mInitialApplication");
        mInitialApplicationField.setAccessible(true);
        mInitialApplicationField.set(mMainThread, delegate);

        //ActivityThread ------>  mAllApplications(ArrayList)   ContextImpl的mMainThread属性
        Field mAllApplicationsField = activityThreadClass.getDeclaredField("mAllApplications");
        mAllApplicationsField.setAccessible(true);
        ArrayList<Application> mApplications = (ArrayList<Application>) mAllApplicationsField.get(mMainThread);
        mApplications.remove(this);
        mApplications.add(delegate);

        //LoadedApk ----->  mApplicaion             ContextImpl的mPackageInfo属性
        Field mPackageInfoField = contextImplClass.getDeclaredField("mPackageInfo");
        mPackageInfoField.setAccessible(true);
        Object mPackageInfo = mPackageInfoField.get(baseContext);


        Class<?> loadedApkClass = Class.forName("android.app.LoadedApk");
        Field mApplicationField = loadedApkClass.getDeclaredField("mApplication");
        mApplicationField.setAccessible(true);
        mApplicationField.set(mPackageInfo, delegate);

        //修改ApplicationInfo  className  LoadedApk
        Field mApplicationInfoField = loadedApkClass.getDeclaredField("mApplicationInfo");
        mApplicationInfoField.setAccessible(true);
        ApplicationInfo mApplicationInfo = (ApplicationInfo) mApplicationInfoField.get(mPackageInfo);
        mApplicationInfo.className = appName;

        delegate.onCreate();
        isBindReal = true;
    }

}

File: proxy_lib/src/main/java/wxgaly/android/proxy_lib/kotlin/ProxyApplication.kt
package wxgaly.android.proxy_lib.kotlin

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.text.TextUtils
import java.io.*
import kotlin.properties.Delegates


/**
 *  wxgaly.android.proxy_lib.
 *
 * @author Created by WXG on 2019-10-28 15:28.
 * @version V1.0
 */
class ProxyApplication : Application() {

    //定义好的加密后的文件的存放路径
    private var app_name by Delegates.notNull<String>()
    private var app_dex_dir by Delegates.notNull<String>()
    private val desCryptUtil =
        DesCryptUtil(DEFAULT_KEY)
    var isBindReal: Boolean = false
    var delegate by Delegates.notNull<Application>()

    override fun attachBaseContext(base: Context?) {
        super.attachBaseContext(base)

        //获取用户填入的metaData
        getMetaData()

        //得到当前apk文件
        val apkFile = File(applicationInfo.sourceDir)

        //把apk解压  这个目录中的内容需要root权限才能使用
        val appDexDir = getDir(app_name + "_" + app_dex_dir, Context.MODE_PRIVATE)

        val appDir = File(appDexDir, "app")
        val dexDir = File(appDir, "dexDir")

        //得到我们需要加载的dex文件
        val dexFiles = mutableListOf<File>()

        //进行解密 （最好做md5文件校验）
        if (!dexDir.exists() || dexDir.list().isEmpty()) {
            //把apk解压到appDir
            ZipUtil.UnZipFolder(apkFile.absolutePath, appDir.absolutePath)
//            Zip.unZip(apkFile,appDir);
            //获取目录下所有的文件
            val files = appDir.listFiles()
            files.forEach { file ->
                val name = file.name
                if (name.endsWith(".dexs") && !TextUtils.equals(name, "classes.dex")) {
                    var fin: FileInputStream? = null
                    var fos: FileOutputStream? = null
                    var bos: ByteArrayOutputStream? = null

                    try {
                        //读取文件内容
                        bos = ByteArrayOutputStream()

                        fin = FileInputStream(file)

                        //解密
                        val decypt = desCryptUtil.decrypt(
                            FileUtil.readFileBytes(
                                file
                            )!!)
                        //写到指定的目录
                        fos = FileOutputStream(file)
                        fos.write(decypt)
                        fos.flush()
                        fos.close()

                        dexFiles.add(file)

                    } catch (e: Exception) {
                        e.printStackTrace()
                    } finally {
                        bos?.close()
                        fin?.close()
                        fos?.close()
                    }
                }
            }

        } else {
            dexDir.listFiles().forEach {
                dexFiles.add(it)
            }
        }

        try {
            loadDex(dexFiles, appDexDir)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onCreate() {
        super.onCreate()
        try {
            bindRealApplication()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun getPackageName(): String {
        if (!TextUtils.isEmpty(app_name)) {
            return ""
        }
        return super.getPackageName()
    }

    override fun createPackageContext(packageName: String?, flags: Int): Context {
        if (TextUtils.isEmpty(app_name)) {
            return super.createPackageContext(packageName, flags)
        }
        try {
            bindRealApplication()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return delegate
    }

    @SuppressLint("PrivateApi", "DiscouragedPrivateApi")
    private fun bindRealApplication() {

        if (isBindReal) {
            return
        }

        if (TextUtils.isEmpty(app_name)) {
            return
        }

        //得到attchBaseContext(context) 传入的上下文 ContextImpl
        val baseContext = baseContext

        //创建用户真实的application  （MyApplication）
        var delegateClass: Class<*>? = Class.forName(app_name)

        delegate = delegateClass!!.newInstance() as Application

        //得到attch()方法
        val attach = Application::class.java.getDeclaredMethod("attach", Context::class.java)
        attach.isAccessible = true
        attach.invoke(delegate, baseContext)

        //获取ContextImpl ----> ,mOuterContext(app);  通过Application的attachBaseContext回调参数获取
        val contextImplClass = Class.forName("android.app.ContextImpl")

        //获取mOuterContext属性
        val mOuterContextField = contextImplClass.getDeclaredField("mOuterContext")
        mOuterContextField.isAccessible = true
        mOuterContextField.set(baseContext, delegate)

        //ActivityThread  ----> mAllApplication(ArrayList)  ContextImpl的mMainThread属性
        val mMainThreadField = contextImplClass.getDeclaredField("mMainThread")
        mMainThreadField.isAccessible = true
        val mMainThread = mMainThreadField.get(baseContext)

        //ActivityThread  ----->  mInitialApplication       ContextImpl的mMainThread属性
        val activityThreadClass = Class.forName("android.app.ActivityThread")
        val mInitialApplicationField = activityThreadClass.getDeclaredField("mInitialApplication")
        mInitialApplicationField.isAccessible = true
        mInitialApplicationField.set(mMainThread, delegate)

        //ActivityThread ------>  mAllApplications(ArrayList)   ContextImpl的mMainThread属性
        val mAllApplicationsField = activityThreadClass.getDeclaredField("mAllApplications")
        mAllApplicationsField.isAccessible = true
        val mApplications = mAllApplicationsField.get(mMainThread) as ArrayList<Application>
        mApplications.remove(this)
        mApplications.add(delegate)

        //LoadedApk ----->  mApplicaion             ContextImpl的mPackageInfo属性
        val mPackageInfoField = contextImplClass.getDeclaredField("mPackageInfo")
        mPackageInfoField.isAccessible = true
        val mPackageInfo = mPackageInfoField.get(baseContext)

        val loadedApkClass = Class.forName("android.app.LoadedApk")
        val mApplicationField = loadedApkClass.getDeclaredField("mApplication")
        mApplicationField.isAccessible = true
        mApplicationField.set(mPackageInfo, delegate)

        //修改ApplicationInfo  className  LoadedApk
        val mApplicationInfoField = loadedApkClass.getDeclaredField("mApplicationInfo")
        mApplicationInfoField.isAccessible = true
        val mApplicationInfo = mApplicationInfoField.get(mPackageInfo) as ApplicationInfo
        mApplicationInfo.className = app_name

        delegate.onCreate()
        isBindReal = true
    }

    @Throws(Exception::class)
    private fun loadDex(dexFiles: List<File>, versionDir: File) {
        //1、获取pathList
        val pathListField = ReflectUtils.getDeclaredField(
            getClassLoader(),
            "pathList"
        )
        val pathList = pathListField?.get(getClassLoader())
        //2、获取数组dexElements
        val dexElementsField =
            ReflectUtils.getDeclaredField(pathList!!, "dexElements")
        val dexElements = dexElementsField?.get(pathList) as Array<Any>
        //3、反射到初始化makePathElements的方法
        val makeDexElements = ReflectUtils.getMethod(
            pathList,
            "makePathElements",
            List::class.java,
            File::class.java,
            List::class.java
        )

        val suppressedException = mutableListOf<IOException>()
        val addElements = makeDexElements!!.invoke(
            pathList,
            dexFiles,
            versionDir,
            suppressedException
        ) as Array<Any>

        val newElements = java.lang.reflect.Array.newInstance(
            dexElements.javaClass.componentType,
            dexElements.size + addElements.size
        ) as Array<Any>

        System.arraycopy(dexElements, 0, newElements, 0, dexElements.size)
        System.arraycopy(addElements, 0, newElements, dexElements.size, addElements.size)

        //替换classloader中的element数组
        dexElementsField.set(pathList, newElements)
    }

    private fun getMetaData() {

        try {
            val applicationInfo = packageManager.getApplicationInfo(
                packageName, PackageManager.GET_META_DATA
            )
            val metaData = applicationInfo.metaData
            if (null != metaData) {
                if (metaData.containsKey(APP_NAME)) {
                    app_name = metaData.getString(APP_NAME)!!
                }
                if (metaData.containsKey(APP_DEX_DIR)) {
                    app_dex_dir = metaData.getString(APP_DEX_DIR)!!
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }


}